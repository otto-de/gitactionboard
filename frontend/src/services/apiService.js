import { clearCookies, fetchAccessToken } from '@/services/authenticationService';

const validate = (res) => {
  if (!res.ok && res.redirected) {
    clearCookies();
    window.location.assign('#/login');
  }

  return res.ok ? Promise.resolve(res) : Promise.reject(res);
};

const isNullOrUndefined = (value) => value === null || value === undefined;

const marshalHeaders = (headers = {}) =>
  Object.keys(headers).reduce(
    (accumulator, key) =>
      isNullOrUndefined(headers[key])
        ? accumulator
        : { ...accumulator, [key]: headers[key] },
    {}
  );

const fetchJsonContent = (url, authToken) => {
  return fetch(url, {
    headers: new Headers(
      marshalHeaders({
        Authorization: authToken,
        Accept: 'application/json'
      })
    )
  })
    .then(validate)
    .then((res) => res.json());
};

export const fetchConfig = () => fetchJsonContent(preparePath('/config'));

export const fetchCctrayJson = () =>
  fetchJsonContent(preparePath('/v1/cctray'), fetchAccessToken());

export const fetchSecretAlerts = () =>
  fetchJsonContent(preparePath('/v1/alerts/secrets'), fetchAccessToken());

export const fetchCodeStandardViolations = () =>
  fetchJsonContent(preparePath('/v1/alerts/code-standard-violations'), fetchAccessToken());

export const fetchRepositoryNames = () =>
  fetchJsonContent(preparePath('/config/repository-names'), fetchAccessToken());

export const fetchWorkflowRunMetrics = (repoName, fromDate, toDate) => {
  const from = fromDate ? fromDate.toISOString() : '';
  const to = toDate ? toDate.toISOString() : '';

  return fetchJsonContent(
    preparePath(`/v1/metrics/workflow-runs/${repoName}?from=${from}&to=${to}`),
    fetchAccessToken());
};

export const authenticate = (username, password) => {
  return fetch(preparePath('/login/basic'), {
    method: 'POST',
    headers: new Headers({ 'Content-Type': 'application/json' }),
    body: JSON.stringify({ username, password })
  }).then(validate);
};

export const preparePath = (url) => `${import.meta.env.VITE_PROXY_TARGET || '.'}${url}`;
