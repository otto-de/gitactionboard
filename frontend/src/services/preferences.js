import storageService from "@/services/storageService";

class Preferences {
  __get__(configKey) {
    return storageService.getItem(configKey);
  }

  __set__(configKey, value) {
    storageService.setItem(configKey, value);
  }

  get showHealthyBuilds() {
    const showHealthBuildConfig = this.__get__("show-healthy-builds");
    return !(showHealthBuildConfig && showHealthBuildConfig === "false");
  }

  set showHealthyBuilds(value) {
    this.__set__("show-healthy-builds", value);
  }

  get disableIdleOptimization() {
    return this.__get__("disable-max-idle-time") === "true";
  }

  set disableIdleOptimization(value) {
    this.__set__("disable-max-idle-time", value);
  }

  get maxIdleTime() {
    const maxIdleTimeConfig = this.__get__("max-idle-time");
    return maxIdleTimeConfig ? parseInt(maxIdleTimeConfig) : 5;
  }

  set maxIdleTime(value) {
    this.__set__("max-idle-time", value);
  }

  get hiddenElements() {
    return JSON.parse(this.__get__("hidden-elements") || "{}");
  }

  set hiddenElements(value) {
    this.__set__("hidden-elements", JSON.stringify(value));
  }
}

export default new Preferences();
