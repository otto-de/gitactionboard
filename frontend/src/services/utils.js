import storageService from "@/services/storageService";

export const setAvailableAuths = (availableAuths) => {
  storageService.setItem("availableAuths", JSON.stringify(availableAuths));
};

export const setGithubSecretsScanMonitoringEnabled = (
  githubSecretsScanMonitoringEnabled
) => {
  storageService.setItem(
    "githubSecretsScanMonitoringEnabled",
    JSON.stringify(githubSecretsScanMonitoringEnabled)
  );
};

export const setGithubCodeScanMonitoringEnabled = (
  githubSecretsScanMonitoringEnabled
) => {
  storageService.setItem(
    "githubCodeScanMonitoringEnabled",
    JSON.stringify(githubSecretsScanMonitoringEnabled)
  );
};

export const getAvailableAuths = () =>
  JSON.parse(storageService.getItem("availableAuths"));

export const getGithubSecretsScanMonitoringEnabled = () =>
  JSON.parse(storageService.getItem("githubSecretsScanMonitoringEnabled")) ||
  false;

export const getGithubCodeScanMonitoringEnabled = () =>
  JSON.parse(storageService.getItem("githubCodeScanMonitoringEnabled")) ||
  false;
