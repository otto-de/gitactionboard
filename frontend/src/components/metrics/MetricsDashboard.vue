<template>
  <v-container
    id="metrics-dashboard"
    fluid
  >
    <DashboardHeader sub-header="Metrics" />
    <v-toolbar
      elevation="4"
      class="mb-2"
    >
      <v-spacer />
      <DateRangePicker
        label="Date Range"
        class="mr-2"
        @date-range-updated="dateRangeUpdated"
      />
      <v-btn
        text="Expand all"
        test-id="expand-all"
        class="text-none"
        rounded
        ripple
        @click="expandAllPanels"
      />
      <v-btn
        text="Collapse all"
        class="text-none"
        test-id="collapse-all"
        rounded
        ripple
        @click="collapseAllPanels"
      />
    </v-toolbar>
    <v-container
      v-if="loading"
      id="spinner-container"
      class="justify-center fill-height spinner-container"
    >
      <Spinner />
    </v-container>
    <v-expansion-panels
      v-else
      v-model="expandedPanels"
      multiple
      static
      test-id="metrics-dashboard-expansion-panels"
    >
      <v-expansion-panel
        v-for="repoName in repositoryNames"
        :key="repoName"
        :title="repoName"
        :value="repoName"
        :test-id="`metrics-dashboard-expansion-panel-${repoName}`"
      >
        <v-expansion-panel-text>
          <RepositoryMetrics
            :repo-name="repoName"
            :from="startDate"
            :to="endDate"
          />
        </v-expansion-panel-text>
      </v-expansion-panel>
    </v-expansion-panels>
  </v-container>
</template>

<script>
import DashboardHeader from '@/components/DashboardHeader.vue';
import { fetchRepositoryNames } from '@/services/apiService';
import Spinner from '@/components/Spinner.vue';
import RepositoryMetrics from '@/components/metrics/RepositoryMetrics.vue';
import DateRangePicker from '@/components/DateRangePicker.vue';

export default {
  name: 'MetricsDashboard',
  components: { DateRangePicker, RepositoryMetrics, Spinner, DashboardHeader },
  data() {
    return {
      loading: true,
      repositoryNames: [],
      startDate: null,
      endDate: null,
      expandedPanels: []
    };
  },
  mounted() {
    fetchRepositoryNames()
      .then(repositoryNames => {
        this.loading = false;
        this.repositoryNames = repositoryNames;
      }).catch(reason => {
        console.error(reason);
        throw reason;
      });
  },
  methods: {
    dateRangeUpdated(from, to) {
      this.startDate = from;
      this.endDate = to;
    },
    expandAllPanels() {
      this.expandedPanels = this.repositoryNames;
    },
    collapseAllPanels() {
      this.expandedPanels = [];
    }
  }
};
</script>

<style scoped>
.spinner-container {
  height: 90vh !important;
}
</style>
