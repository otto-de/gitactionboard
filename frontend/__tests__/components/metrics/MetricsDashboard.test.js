import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import Spinner from '@/components/Spinner.vue';

import { mountWithWrapper, promiseWithResolvers, requestAnimationFrameAsPromise } from '../../test-utils';
import flushPromises from 'flush-promises';
import MetricsDashboard from '@/components/metrics/MetricsDashboard.vue';
import { fetchRepositoryNames, fetchWorkflowRunMetrics } from '@/services/apiService';
import preferences from '@/services/preferences';
import { getVersion } from '@/services/utils';

vi.mock('@/services/apiService');
vi.mock('@/services/utils');

describe('MetricsDashboard', () => {
  const FIXED_SYSTEM_TIME = '2024-08-22T12:00:00.000Z';
  const FROM = new Date('2024-08-14T18:30:00.000Z');
  const TO = new Date('2024-08-22T18:29:59.999Z');

  beforeEach(() => {
    getVersion.mockReturnValueOnce('5.0.0');
    vi.spyOn(preferences, 'theme', 'get').mockReturnValueOnce('light');
    vi.useFakeTimers();
    vi.setSystemTime(Date.parse(FIXED_SYSTEM_TIME));
  });

  afterEach(() => {
    vi.clearAllMocks();
    vi.useRealTimers();
  });

  const repoName1 = 'hello_world';
  const repoName2 = 'test';

  const mockSuccessfulWorkflowRunMetrics = () => {
    fetchWorkflowRunMetrics.mockImplementation((repoName) => {
      return Promise.resolve({
        test: [
          {
            workflowName: 'test',
            repoName,
            conclusion: 'FAILURE',
            runAttempt: 1,
            startedAt: '2024-09-06T06:06:04.000Z',
            completedAt: '2024-09-06T06:09:44.000Z',
            triggeredEvent: 'schedule'
          },
          {
            workflowName: 'test',
            repoName,
            conclusion: 'FAILURE',
            runAttempt: 1,
            startedAt: '2024-09-05T11:18:38.000Z',
            completedAt: '2024-09-05T11:22:06.000Z',
            triggeredEvent: 'push'
          }
        ]
      });
    });
  };

  it('should display spinner when component is loading', async () => {
    fetchRepositoryNames.mockReturnValue(promiseWithResolvers().promise);

    const wrapper = mountWithWrapper(MetricsDashboard);

    await vi.waitUntil(() => wrapper.findComponent(Spinner).exists());

    vi.useRealTimers();

    await requestAnimationFrameAsPromise();

    expect(wrapper.findComponent(MetricsDashboard).html()).toMatchSnapshot();
  });

  it('should display repository metrics expansion panels once the component is loaded', async () => {
    const repositoryNames = [repoName1, repoName2];
    fetchRepositoryNames.mockResolvedValueOnce(repositoryNames);

    const wrapper = mountWithWrapper(MetricsDashboard);

    await vi.waitUntil(() => wrapper.find("[test-id='metrics-dashboard-expansion-panels']").exists());

    expect(wrapper.findComponent(Spinner).exists()).toBeFalsy();

    repositoryNames.forEach(repositoryName =>
      expect(wrapper.find(`[test-id='metrics-dashboard-expansion-panel-${repositoryName}']`).exists())
        .toBeTruthy());
  });

  it('should display repository metrics when clicks on expansion icon', async () => {
    const repositoryNames = [repoName1, repoName2];
    fetchRepositoryNames.mockResolvedValueOnce(repositoryNames);

    mockSuccessfulWorkflowRunMetrics();

    const wrapper = mountWithWrapper(MetricsDashboard);

    await vi.waitUntil(() => wrapper.find(`[test-id='metrics-dashboard-expansion-panel-${repoName1}']`).exists());

    const repoName1ExpansionPanelWrapper = wrapper.find(`[test-id='metrics-dashboard-expansion-panel-${repoName1}']`);
    await repoName1ExpansionPanelWrapper.find('.mdi-chevron-down').trigger('click');
    await vi.waitUntil(() => wrapper.find(`#${repoName1}-metrics`).exists());

    expect(wrapper.find("[test-id='metrics-dashboard-expansion-panels']")).toMatchSnapshot();
    expect(fetchWorkflowRunMetrics).toHaveBeenCalledOnce();
    expect(fetchWorkflowRunMetrics)
      .toHaveBeenCalledWith(repoName1, FROM,
        TO);
    expect(wrapper.find(`#${repoName2}-metrics`).exists()).toBeFalsy();
  });

  it('should hide repository metrics when clicks on collapse icon', async () => {
    const repositoryNames = [repoName1, repoName2];
    fetchRepositoryNames.mockResolvedValueOnce(repositoryNames);

    mockSuccessfulWorkflowRunMetrics();

    const wrapper = mountWithWrapper(MetricsDashboard);

    await vi.waitUntil(() => wrapper.find(`[test-id='metrics-dashboard-expansion-panel-${repoName1}']`).exists());

    await wrapper.find(`[test-id='metrics-dashboard-expansion-panel-${repoName1}']`)
      .find('.mdi-chevron-down').trigger('click');

    await vi.waitUntil(() => wrapper.find(`#${repoName1}-metrics`).exists());

    await wrapper.find(`[test-id='metrics-dashboard-expansion-panel-${repoName1}']`)
      .find('.mdi-chevron-up').trigger('click');
    await wrapper.vm.$nextTick();
    await flushPromises();
    await wrapper.vm.$nextTick();

    expect(wrapper.find(`[test-id='metrics-dashboard-expansion-panel-${repoName1}']`)
      .find('.mdi-chevron-up').exists()).toBeFalsy();
    // TODO: Check RepositoryMetrics components gets unmounted
  });

  it('should display all repository metrics when clicks on expand all button', async () => {
    const repositoryNames = [repoName1, repoName2];
    fetchRepositoryNames.mockResolvedValueOnce(repositoryNames);

    mockSuccessfulWorkflowRunMetrics();

    const wrapper = mountWithWrapper(MetricsDashboard);

    await vi.waitUntil(() => wrapper.find("[test-id='metrics-dashboard-expansion-panels']").exists());

    await wrapper.find("[test-id='expand-all']").trigger('click');

    await vi.waitUntil(() => wrapper.find(`#${repoName1}-metrics`).exists());
    await vi.waitUntil(() => wrapper.find(`#${repoName2}-metrics`).exists());

    expect(fetchWorkflowRunMetrics).toHaveBeenCalledTimes(2);
    expect(fetchWorkflowRunMetrics).toHaveBeenCalledWith(repoName1,
      FROM, TO);
    expect(fetchWorkflowRunMetrics).toHaveBeenCalledWith(repoName2,
      FROM, TO);
  });

  it('should collapse all repository metrics when clicks on collapse all button', async () => {
    const repositoryNames = [repoName1, repoName2];
    fetchRepositoryNames.mockResolvedValueOnce(repositoryNames);

    mockSuccessfulWorkflowRunMetrics();

    const wrapper = mountWithWrapper(MetricsDashboard);

    await vi.waitUntil(() => wrapper.find("[test-id='metrics-dashboard-expansion-panels']").exists());

    await wrapper.find("[test-id='expand-all']").trigger('click');
    await wrapper.vm.$nextTick();
    await flushPromises();

    await vi.waitUntil(() => wrapper.find(`#${repoName1}-metrics`).exists());
    await vi.waitUntil(() => wrapper.find(`#${repoName2}-metrics`).exists());

    await wrapper.find("[test-id='collapse-all']").trigger('click');
    await wrapper.vm.$nextTick();
    await flushPromises();

    expect(wrapper.find(`[test-id='metrics-dashboard-expansion-panel-${repoName1}']`)
      .find('.mdi-chevron-up').exists()).toBeFalsy();
    expect(wrapper.find(`[test-id='metrics-dashboard-expansion-panel-${repoName2}']`)
      .find('.mdi-chevron-up').exists()).toBeFalsy();
  });
});
