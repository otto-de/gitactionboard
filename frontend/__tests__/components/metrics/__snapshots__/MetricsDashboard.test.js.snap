// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MetricsDashboard > should display repository metrics when clicks on expansion icon 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <div
    class="v-expansion-panels v-theme--light v-expansion-panels--variant-default"
    data-v-a30323f4=""
    test-id="metrics-dashboard-expansion-panels"
  >
    
    
    <div
      class="v-expansion-panel v-expansion-panel--active"
      data-v-a30323f4=""
      test-id="metrics-dashboard-expansion-panel-hello_world"
    >
      <div
        class="v-expansion-panel__shadow"
      />
      
      <button
        aria-expanded="true"
        class="v-expansion-panel-title v-expansion-panel-title--active v-expansion-panel-title--static"
        type="button"
      >
        <span
          class="v-expansion-panel-title__overlay"
        />
        
        hello_world
        
        
        <span
          class="v-expansion-panel-title__icon"
        >
          <i
            aria-hidden="true"
            class="mdi-chevron-up mdi v-icon notranslate v-theme--light v-icon--size-default"
          />
        </span>
        
      </button>
      <!---->
      
      <transition-stub
        appear="false"
        css="true"
        data-v-a30323f4=""
        name="expand-transition"
        persisted="false"
      >
        <div
          class="v-expansion-panel-text"
          style=""
        >
          <div
            class="v-expansion-panel-text__wrapper"
          >
            
            <div
              class="v-container v-container--fluid v-locale--is-ltr"
              data-v-a30323f4=""
              id="hello_world-metrics"
            >
              <div
                class="v-row v-row--dense"
              >
                <div
                  class="v-col v-col-3"
                >
                  
                  <div
                    class="v-card v-theme--light v-card--density-default v-card--variant-elevated border-thin"
                  >
                    <!---->
                    <div
                      class="v-card__loader"
                    >
                      <div
                        aria-hidden="true"
                        aria-valuemax="100"
                        aria-valuemin="0"
                        class="v-progress-linear v-theme--light v-locale--is-ltr"
                        role="progressbar"
                        style="top: 0px; height: 0px; --v-progress-linear-height: 2px;"
                      >
                        <!---->
                        <div
                          class="v-progress-linear__background"
                        />
                        <div
                          class="v-progress-linear__buffer"
                          style="width: 0%;"
                        />
                        <transition-stub
                          appear="false"
                          css="true"
                          name="fade-transition"
                          persisted="false"
                        >
                          <div
                            class="v-progress-linear__indeterminate"
                          >
                            
                            <div
                              class="v-progress-linear__indeterminate long"
                            />
                            <div
                              class="v-progress-linear__indeterminate short"
                            />
                            
                          </div>
                        </transition-stub>
                        <!---->
                      </div>
                    </div>
                    <!---->
                    <!---->
                    
                    <div
                      class="v-card-subtitle"
                    >
                      <div
                        class="v-row align-center justify-space-between"
                      >
                        <div
                          class="v-col v-col-auto"
                        >
                          Workflow Run Frequency
                        </div>
                        <div
                          class="v-col v-col-auto"
                        >
                          <button
                            aria-label="Expand"
                            class="v-btn v-btn--elevated v-btn--icon v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-elevated"
                            type="button"
                          >
                            
                            <span
                              class="v-btn__overlay"
                            />
                            <span
                              class="v-btn__underlay"
                            />
                            
                            <!---->
                            <span
                              class="v-btn__content"
                              data-no-activator=""
                            >
                              <i
                                aria-hidden="true"
                                class="mdi-arrow-expand mdi v-icon notranslate v-theme--light v-icon--size-default"
                              />
                            </span>
                            <!---->
                            <!---->
                          </button>
                        </div>
                      </div>
                    </div>
                    <canvas
                      height="0"
                      id="hello_world_Workflow_Run_Frequency_grid"
                      role="img"
                      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
                      width="0"
                    >
                      <p>
                        
                      </p>
                    </canvas>
                    
                    <!---->
                    
                    <!---->
                    <span
                      class="v-card__underlay"
                    />
                    
                  </div>
                  
                  <!---->
                  <!---->
                  
                  
                </div>
                <div
                  class="v-col v-col-3"
                >
                  
                  <div
                    class="v-card v-theme--light v-card--density-default v-card--variant-elevated border-thin"
                  >
                    <!---->
                    <div
                      class="v-card__loader"
                    >
                      <div
                        aria-hidden="true"
                        aria-valuemax="100"
                        aria-valuemin="0"
                        class="v-progress-linear v-theme--light v-locale--is-ltr"
                        role="progressbar"
                        style="top: 0px; height: 0px; --v-progress-linear-height: 2px;"
                      >
                        <!---->
                        <div
                          class="v-progress-linear__background"
                        />
                        <div
                          class="v-progress-linear__buffer"
                          style="width: 0%;"
                        />
                        <transition-stub
                          appear="false"
                          css="true"
                          name="fade-transition"
                          persisted="false"
                        >
                          <div
                            class="v-progress-linear__indeterminate"
                          >
                            
                            <div
                              class="v-progress-linear__indeterminate long"
                            />
                            <div
                              class="v-progress-linear__indeterminate short"
                            />
                            
                          </div>
                        </transition-stub>
                        <!---->
                      </div>
                    </div>
                    <!---->
                    <!---->
                    
                    <div
                      class="v-card-subtitle"
                    >
                      <div
                        class="v-row align-center justify-space-between"
                      >
                        <div
                          class="v-col v-col-auto"
                        >
                          Workflow Run Failure Rate
                        </div>
                        <div
                          class="v-col v-col-auto"
                        >
                          <button
                            aria-label="Expand"
                            class="v-btn v-btn--elevated v-btn--icon v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-elevated"
                            type="button"
                          >
                            
                            <span
                              class="v-btn__overlay"
                            />
                            <span
                              class="v-btn__underlay"
                            />
                            
                            <!---->
                            <span
                              class="v-btn__content"
                              data-no-activator=""
                            >
                              <i
                                aria-hidden="true"
                                class="mdi-arrow-expand mdi v-icon notranslate v-theme--light v-icon--size-default"
                              />
                            </span>
                            <!---->
                            <!---->
                          </button>
                        </div>
                      </div>
                    </div>
                    <canvas
                      height="0"
                      id="hello_world_Workflow_Run_Failure_Rate_grid"
                      role="img"
                      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
                      width="0"
                    >
                      <p>
                        
                      </p>
                    </canvas>
                    
                    <!---->
                    
                    <!---->
                    <span
                      class="v-card__underlay"
                    />
                    
                  </div>
                  
                  <!---->
                  <!---->
                  
                  
                </div>
                <div
                  class="v-col v-col-3"
                >
                  
                  <div
                    class="v-card v-theme--light v-card--density-default v-card--variant-elevated border-thin"
                  >
                    <!---->
                    <div
                      class="v-card__loader"
                    >
                      <div
                        aria-hidden="true"
                        aria-valuemax="100"
                        aria-valuemin="0"
                        class="v-progress-linear v-theme--light v-locale--is-ltr"
                        role="progressbar"
                        style="top: 0px; height: 0px; --v-progress-linear-height: 2px;"
                      >
                        <!---->
                        <div
                          class="v-progress-linear__background"
                        />
                        <div
                          class="v-progress-linear__buffer"
                          style="width: 0%;"
                        />
                        <transition-stub
                          appear="false"
                          css="true"
                          name="fade-transition"
                          persisted="false"
                        >
                          <div
                            class="v-progress-linear__indeterminate"
                          >
                            
                            <div
                              class="v-progress-linear__indeterminate long"
                            />
                            <div
                              class="v-progress-linear__indeterminate short"
                            />
                            
                          </div>
                        </transition-stub>
                        <!---->
                      </div>
                    </div>
                    <!---->
                    <!---->
                    
                    <div
                      class="v-card-subtitle"
                    >
                      <div
                        class="v-row align-center justify-space-between"
                      >
                        <div
                          class="v-col v-col-auto"
                        >
                          Workflow Re-trigger Rate
                        </div>
                        <div
                          class="v-col v-col-auto"
                        >
                          <button
                            aria-label="Expand"
                            class="v-btn v-btn--elevated v-btn--icon v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-elevated"
                            type="button"
                          >
                            
                            <span
                              class="v-btn__overlay"
                            />
                            <span
                              class="v-btn__underlay"
                            />
                            
                            <!---->
                            <span
                              class="v-btn__content"
                              data-no-activator=""
                            >
                              <i
                                aria-hidden="true"
                                class="mdi-arrow-expand mdi v-icon notranslate v-theme--light v-icon--size-default"
                              />
                            </span>
                            <!---->
                            <!---->
                          </button>
                        </div>
                      </div>
                    </div>
                    <canvas
                      height="0"
                      id="hello_world_Workflow_Re-trigger_Rate_grid"
                      role="img"
                      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
                      width="0"
                    >
                      <p>
                        
                      </p>
                    </canvas>
                    
                    <!---->
                    
                    <!---->
                    <span
                      class="v-card__underlay"
                    />
                    
                  </div>
                  
                  <!---->
                  <!---->
                  
                  
                </div>
                <div
                  class="v-col v-col-3"
                >
                  
                  <div
                    class="v-card v-theme--light v-card--density-default v-card--variant-elevated border-thin"
                  >
                    <!---->
                    <div
                      class="v-card__loader"
                    >
                      <div
                        aria-hidden="true"
                        aria-valuemax="100"
                        aria-valuemin="0"
                        class="v-progress-linear v-theme--light v-locale--is-ltr"
                        role="progressbar"
                        style="top: 0px; height: 0px; --v-progress-linear-height: 2px;"
                      >
                        <!---->
                        <div
                          class="v-progress-linear__background"
                        />
                        <div
                          class="v-progress-linear__buffer"
                          style="width: 0%;"
                        />
                        <transition-stub
                          appear="false"
                          css="true"
                          name="fade-transition"
                          persisted="false"
                        >
                          <div
                            class="v-progress-linear__indeterminate"
                          >
                            
                            <div
                              class="v-progress-linear__indeterminate long"
                            />
                            <div
                              class="v-progress-linear__indeterminate short"
                            />
                            
                          </div>
                        </transition-stub>
                        <!---->
                      </div>
                    </div>
                    <!---->
                    <!---->
                    
                    <div
                      class="v-card-subtitle"
                    >
                      <div
                        class="v-row align-center justify-space-between"
                      >
                        <div
                          class="v-col v-col-auto"
                        >
                          Average Workflow Run Time (Minutes)
                        </div>
                        <div
                          class="v-col v-col-auto"
                        >
                          <button
                            aria-label="Expand"
                            class="v-btn v-btn--elevated v-btn--icon v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-elevated"
                            type="button"
                          >
                            
                            <span
                              class="v-btn__overlay"
                            />
                            <span
                              class="v-btn__underlay"
                            />
                            
                            <!---->
                            <span
                              class="v-btn__content"
                              data-no-activator=""
                            >
                              <i
                                aria-hidden="true"
                                class="mdi-arrow-expand mdi v-icon notranslate v-theme--light v-icon--size-default"
                              />
                            </span>
                            <!---->
                            <!---->
                          </button>
                        </div>
                      </div>
                    </div>
                    <canvas
                      height="0"
                      id="hello_world_Average_Workflow_Run_Time_(Minutes)_grid"
                      role="img"
                      style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
                      width="0"
                    >
                      <p>
                        
                      </p>
                    </canvas>
                    
                    <!---->
                    
                    <!---->
                    <span
                      class="v-card__underlay"
                    />
                    
                  </div>
                  
                  <!---->
                  <!---->
                  
                  
                </div>
              </div>
            </div>
            
          </div>
        </div>
      </transition-stub>
      
      
    </div>
    <div
      class="v-expansion-panel v-expansion-panel--after-active"
      data-v-a30323f4=""
      test-id="metrics-dashboard-expansion-panel-test"
    >
      <div
        class="v-expansion-panel__shadow"
      />
      
      <button
        aria-expanded="false"
        class="v-expansion-panel-title v-expansion-panel-title--static"
        type="button"
      >
        <span
          class="v-expansion-panel-title__overlay"
        />
        
        test
        
        
        <span
          class="v-expansion-panel-title__icon"
        >
          <i
            aria-hidden="true"
            class="mdi-chevron-down mdi v-icon notranslate v-theme--light v-icon--size-default"
          />
        </span>
        
      </button>
      <!---->
      
      <transition-stub
        appear="false"
        css="true"
        data-v-a30323f4=""
        name="expand-transition"
        persisted="false"
      >
        <div
          class="v-expansion-panel-text"
          style="display: none;"
        >
          <!---->
        </div>
      </transition-stub>
      
      
    </div>
    
    
  </div>,
}
`;

exports[`MetricsDashboard > should display spinner when component is loading 1`] = `
<div class="v-container v-container--fluid v-locale--is-ltr" id="metrics-dashboard">
  <header class="v-toolbar v-toolbar--density-default elevation-5 v-theme--light v-locale--is-ltr v-app-bar" style="top: 0px; z-index: 1004; transform: translateY(0px); position: fixed; left: 0px; width: calc(100% - 0px - 0px);">
    <!---->
    <div class="v-toolbar__content" style="height: 64px;">
      <!---->
      <!---->
      <div class="v-row v-row--no-gutters">
        <div class="v-col v-col-10">
          <div class="v-toolbar-title header-title text-center font-weight-bold pl-10 v-app-bar-title">
            <div class="v-toolbar-title__placeholder">
              <!---->
              <span>Gitaction Board</span>
              <span class="version">(v5.0.0)</span>
            </div>
          </div>
        </div>
        <div class="v-col v-col-2 d-inline-flex">
          <div class="v-toolbar-title pr-4 text-right font-weight-bold align-center fill-height d-inline-flex flex-row-reverse mt-3 v-app-bar-title" id="sub-header">
            <div class="v-toolbar-title__placeholder">
              <!---->
              Metrics
            </div>
          </div>
        </div>
      </div>
      <!---->
    </div>
    <transition-stub appear="false" css="true" name="expand-transition" persisted="false">
      <!---->
    </transition-stub>
  </header>
  <header class="v-toolbar v-toolbar--density-default elevation-4 v-theme--light v-locale--is-ltr mb-2">
    <!---->
    <div class="v-toolbar__content" style="height: 64px;">
      <!---->
      <!---->
      <div class="v-spacer"></div>
      <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-input--dirty v-text-field mr-2" style="max-width: 350px;">
        <div class="v-input__prepend">
          <!---->
          <i aria-hidden="false" aria-label="Date Range prepended action" class="mdi-calendar mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable" role="button" tabindex="0"></i>
        </div>
        <div class="v-input__control">
          <div aria-expanded="false" aria-haspopup="menu" aria-owns="v-menu-3" class="v-field v-field--active v-field--center-affix v-field--dirty v-field--variant-filled v-theme--light v-locale--is-ltr">
            <div class="v-field__overlay"></div>
            <div class="v-field__loader">
              <div aria-hidden="true" aria-valuemax="100" aria-valuemin="0" class="v-progress-linear v-theme--light v-locale--is-ltr" role="progressbar" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;">
                <!---->
                <div class="v-progress-linear__background"></div>
                <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                <transition-stub appear="false" css="true" name="fade-transition" persisted="false">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <div class="v-field__field" data-no-activator>
              <label aria-hidden="true" class="v-label v-field-label v-field-label--floating" for="input-1">
                <!---->
                Date Range
              </label>
              <label class="v-label v-field-label" for="input-1">
                <!---->
                Date Range
              </label>
              <!---->
              <div class="v-field__input" data-no-activator>
                <!---->
                <!---->
                <!---->
                <input aria-describedby="input-1-messages" id="input-1" placeholder="mm/dd/yyyy" size="1" test-id="vuetify-date-range-picker" type="text" value="08/15/2024 - 08/22/2024">
              </div>
              <!---->
            </div>
            <!---->
            <!---->
            <div class="v-field__outline">
              <!---->
              <!---->
            </div>
          </div>
        </div>
        <!---->
        <div class="v-input__details">
          <transition-group-stub appear="false" aria-live="polite" class="v-messages" css="true" id="input-1-messages" name="slide-y-transition" persisted="false" role="alert" tag="div">
            <!---->
          </transition-group-stub>
          <!---->
        </div>
      </div>
      <button class="v-btn v-theme--light v-btn--density-default v-btn--rounded v-btn--size-default v-btn--variant-text text-none" test-id="expand-all" type="button">
        <span class="v-btn__overlay"></span>
        <span class="v-btn__underlay"></span>
        <!---->
        <span class="v-btn__content" data-no-activator>Expand all</span>
        <!---->
        <!---->
      </button>
      <button class="v-btn v-theme--light v-btn--density-default v-btn--rounded v-btn--size-default v-btn--variant-text text-none" test-id="collapse-all" type="button">
        <span class="v-btn__overlay"></span>
        <span class="v-btn__underlay"></span>
        <!---->
        <span class="v-btn__content" data-no-activator>Collapse all</span>
        <!---->
        <!---->
      </button>
      <!---->
    </div>
    <transition-stub appear="false" css="true" name="expand-transition" persisted="false">
      <!---->
    </transition-stub>
  </header>
  <div class="v-container v-locale--is-ltr justify-center fill-height spinner-container" id="spinner-container">
    <div aria-valuemax="100" aria-valuemin="0" class="v-progress-circular v-progress-circular--indeterminate v-theme--light" role="progressbar" style="width: 128px; height: 128px; color: rgb(95, 141, 119); caret-color: #5f8d77;">
      <svg style="transform: rotate(calc(-90deg + 0deg));" viewBox="0 0 44.13793103448276 44.13793103448276" xmlns="http://www.w3.org/2000/svg">
        <circle class="v-progress-circular__underlay" cx="50%" cy="50%" fill="transparent" r="20" stroke-dasharray="125.66370614359172" stroke-dashoffset="0" stroke-width="4.137931034482759" />
        <circle class="v-progress-circular__overlay" cx="50%" cy="50%" fill="transparent" r="20" stroke-dasharray="125.66370614359172" stroke-dashoffset="125.66370614359172px" stroke-width="4.137931034482759" />
      </svg>
      <!---->
    </div>
  </div>
</div>
`;
